geozero/src/csv/csv_reader.rs

```diff
@@ -0,0 +1,417 @@
+    }
+}
+
+pub struct CsvString {
+    csv_text: String,
+    geometry_column_name: String,
+}
+
+impl CsvString {
+    pub fn new(geometry_column_name: &str, csv_text: String) -> Self {
+        Self {
+            csv_text,
+            geometry_column_name: geometry_column_name.to_string(),
+        }
+    }
+}
+
+impl GeozeroDatasource for CsvString {
+    fn process<P: FeatureProcessor>(&mut self, processor: &mut P) -> Result<()> {
+        process_csv_features(
+            self.csv_text.as_bytes(),
+            processor,
+            &self.geometry_column_name,
+        )
+    }
+}
+
+impl GeozeroGeometry for CsvString {
+    fn process_geom<P: GeomProcessor>(&self, processor: &mut P) -> Result<()> {
+        process_csv_geom(
+            self.csv_text.as_bytes(),
+            processor,
+            &self.geometry_column_name,
+        )
+    }
+}
+
+pub struct CsvReader<R: Read> {
+    inner: R,
+    geometry_column_name: String,
+}
+
+impl<R: Read> CsvReader<R> {
+    pub fn new(geometry_column_name: &str, inner: R) -> Self {
+        Self {
+            inner,
+            geometry_column_name: geometry_column_name.to_string(),
+        }
+    }
+}
+
+impl<R: Read> GeozeroDatasource for CsvReader<R> {
+    fn process<P: FeatureProcessor>(&mut self, processor: &mut P) -> Result<()> {
+        process_csv_features(&mut self.inner, processor, &self.geometry_column_name)
+    }
+}
+
+impl<R: Read + Clone> GeozeroGeometry for CsvReader<R> {
+    fn process_geom<P: GeomProcessor>(&self, processor: &mut P) -> Result<()> {
+        process_csv_geom(self.inner.clone(), processor, &self.geometry_column_name)
+    }
+}
+
+pub fn process_csv_geom(
+    input: impl Read,
+    processor: &mut impl GeomProcessor,
+    geometry_column: &str,
+) -> Result<()> {
+    let mut reader = csv::Reader::from_reader(input);
+    let headers = reader.headers()?.clone();
+
+    let geometry_idx = headers
+        .iter()
+        .position(|f| f == geometry_column)
+        .ok_or(GeozeroError::ColumnNotFound)?;
+
+    // REVIEW: We don't know the size of the geometry collection, so we lie and say it's 0.
+    // Alternatively we'd have to re-traverse the file an extra time just to count or hold
+    // the entire thing in memory, both seem undesirable if the current behavior's only downside
+    // is that it doesn't pre-allocate the geometry collection Vec.
+    processor.geometrycollection_begin(0, 0)?;
+
+    for (record_idx, record) in reader.into_records().enumerate() {
+        let record = record?;
+        let geometry_field = record
+            .get(geometry_idx)
+            .ok_or(GeozeroError::ColumnNotFound)?;
+        use std::str::FromStr;
+        let wkt = wkt::Wkt::from_str(&geometry_field)
+            .map_err(|e| GeozeroError::Geometry(e.to_string()))?;
+        crate::wkt::wkt_reader::process_wkt_geom_n(&wkt.item, record_idx, processor).map_err(
+            |e| {
+                log::warn!("invalid WKT: {}", geometry_field);
```

Comment:
I would expect skipping this geometry error, when a warning is issued? Is this a debugging leftover?

Reply:
It was intentional. I thought it made sense to fail when encountering invalid data, but I wanted to at least offer the user a way to track down the offending row.

Is there other precedence for how to handle partially readable data in the library? I'm a bit nervous about dropping data, potentially silently if the use hasn't enabled logging or maybe isn't even using the library in a text/terminal context.

Reply:
I just pushed an update with a better error message - matching the similar method in the FeatureProcessor impl below.